### PR TITLE
Block reward lock with signed mining

### PR DESCRIPTION
**Pull request by sorpaas**
### Overview

This PR implements miner reward lock together with signed mining. Note that miner reward lock is still under discussion and is not necessarily accepted by Kulupu governance.

The theory is that a miner provides security at a specific point-in-time for Kulupu (by producing a block). However, that specific point-in-time security is expected to have a lasting effect for the blockchain for a certain period of time.

Miner reward minted has locks together with it. First, the voluntary taxation is deducted. After this, at least 1 KLP will always be without locks, so that the miner can always have money to pay for transaction fees. For the remainder of the reward, it has a total lock of 100 days, and the locks expire every 10 days releasing 10% of the reward.

The locked reward can be used to participate in governance, but cannot be transferred.

Note that miner reward lock cannot be used alone -- otherwise the miner can just mine to a "trusted third-party" and convert it to a future asset instead. Here we combine mining reward lock with **signed mining** to deal with this issue. See https://corepaper.org/mining/signed/

### Public testnet `breaknet4`

To test this branch, we created a new public testnet called `breaknet3`. Instruction on how to use this are as below.

First, use `subkey` to generate a new wallet:

```
subkey --sr25519 --network kulupu generate
```

Remember your seed phrase and public key. Then check out this branch, update Substrate vendor:

```
git checkout sp-reward-lock
git submodule update
```

Run the node as below:

```
cargo run --release -- --chain breaknet4 --validator --author <your public key> --register-mining-key "<your seed phrase>"
```

For subsequent runs, you can remove `--register-mining-key "<your seed phrase>"` as the mining key is already registered.

### CODE REVIEW

File: frame/rewards/src/lib.rs [hunk 1]
```diff
@@ -137,14 +185,41 @@ decl_module! {
...
-				drop(T::Currency::deposit_creating(&author, miner));
+				let miner_pref = <Self as Store>::AuthorLockPref::get().unwrap_or_default();
+				let miner_total = reward.saturating_sub(tax);
+				let (miner_reward, lock_days) = match miner_pref {
+					LockPref::None => (miner_total / 3.into(), 0),
+					LockPref::Single => (miner_total / 3.into() + miner_total / 12.into(), 8),
+					LockPref::Double => (miner_total / 2.into(), 16),
+					LockPref::Triple => (miner_total / 2.into() + miner_total / 12.into(), 32),
+					LockPref::Quadruple => (miner_total / 3.into() + miner_total / 3.into(), 64),
+					LockPref::Quintuple => (miner_total / 2.into() + miner_total / 2.into(), 128),
```
Comment (shawntabrizi) [inline frame/rewards/src/lib.rs:196]: 1/2 + 1/2 = 1?

**Comment by shawntabrizi** [comment]
Let me know when this is ready to progress and I can help write tests

**Comment by sorpaas** [comment]
Should have marked this "underdiscussion" -- we haven't yet preliminarily agreed on how the miner reward lock should be implemented.

**Comment by emielvanderhoek** [comment]
I urgently need to wrap up some other work to clear my schedule to work on this. 

Currently my contributions are limited to Discord posts. I plan to digest that into a more concrete overview and proposal in the upcoming week.

File: src/chain_spec.rs [hunk 2]
```diff
@@ -67,11 +67,64 @@ pub fn local_testnet_config() -> ChainSpec {
 	)
 }
 
+pub fn breaknet4_config() -> ChainSpec {
```
Comment (shawntabrizi) [inline src/chain_spec.rs:70]: are you interested to remove this before master merge?

Comment (sorpaas) [reply src/chain_spec.rs:70]: I want to keep it on master at least for the next release. Most users only download release software, so we can get more people testing `breaknet4` before retiring it.

**Review by shawntabrizi — APPROVED** [review]
I walked through all the changes since my last approval. Nothing jumps out to me as out of ordinary or having potential issue, but certainly I am much more unfamiliar with modifying these parts of substrate than you.

That being said, things look reasonable to me

**Pull request merged**